Mass flow rated weighted postprocessor

snschune opened the pull request:
## Reason

To compute averages weighted with mass flux over over boundaries.

## Design

New postprocessor.

## Impact

New capability.

Commit:
Implement MfrWeightedAverage postprocessor (#24676)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1587518/) on b0fad04 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/24680/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/24680/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 4278fbb79d3acc746326873e9f53539d9aba697f`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1587602/) on 11bda11 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24680/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1587643/) on 11bda11 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>4278fb</th>
      <th colspan=4>#24680 11bda1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.40%</td>
      <td>85.40%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88878</td>
      <td>88882</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15197</td>
      <td>15193</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24680/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24680/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>4278fb</th>
      <th colspan=4>#24680 11bda1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.86%</td>
      <td>83.90%</td>
      <td>+0.04%</td>
      <td>97.30%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10802</td>
      <td>10838</td>
      <td>+36</td>
      <td>36</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2079</td>
      <td>2080</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24680/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24680/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24680/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/MfrWeightedAverage.md:
```diff
@@ -0,0 +1,34 @@
+# VolumetricFlowRate
```
update

GiudGiud commented on modules/navier_stokes/doc/content/source/postprocessors/MfrWeightedAverage.md:
```diff
@@ -0,0 +1,34 @@
+# VolumetricFlowRate
+
+!syntax description /Postprocessors/MfrWeightedAverage
```
Mfr stands for Mass Flow Rate and yet this is Mass Flux weighted.

so let s MassFluxWeightedAverage, spelled out

GiudGiud replied:
actually it s also a flow rate

so it's MassFluxWeightedFlowRate ?

GiudGiud commented on modules/navier_stokes/include/postprocessors/MfrWeightedAverage.h:
```diff
@@ -0,0 +1,42 @@
+class SinglePhaseFluidProperties;
+
+/**
+ * This postprocessor computes the mass-flow rate weighted average of a flow
```
```suggestion
 * This postprocessor computes the mass-flux weighted average of a flow
```

GiudGiud commented on modules/navier_stokes/src/postprocessors/MfrWeightedAverage.C:
```diff
@@ -0,0 +1,89 @@
+  InputParameters params = VolumetricFlowRate::validParams();
+  params.addRequiredParam<MooseFunctorName>("density",
+                                            "Provide a functor that returns the density.");
+  params.addClassDescription("Computes the mass flow rate weighted average of the quantity "
```
```suggestion
  params.addClassDescription("Computes the mass flux weighted average of the quantity "
```

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/tests:
```diff
@@ -128,6 +128,15 @@
     []
   []
 
+  [pinsfv_mfr_weighted]
```
```suggestion
  [pinsfv_mass_flux_weighted]
```

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/tests:
```diff
@@ -128,6 +128,15 @@
+    type = CSVDiff
+    input = mfr_weighted_pp.i
+    csvdiff = 'mfr_weighted_pp_out.csv'
+    requirement = 'The system shall compute mass-flow rate averaged quantities over boundaries.'
```
```suggestion
    requirement = 'The system shall compute mass-flux weighted averages of quantities over boundaries.'
```

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/tests:
```diff
@@ -128,6 +128,15 @@
+    input = mfr_weighted_pp.i
+    csvdiff = 'mfr_weighted_pp_out.csv'
+    requirement = 'The system shall compute mass-flow rate averaged quantities over boundaries.'
+    design = 'MfrWeightedAverage.md'
```
```suggestion
    design = 'MassFluxWeightedAverage.md'
```

GiudGiud reviewed (changes requested):
quick look

I dont like the proliferation of abbreviations in general.
For us it s trivial, for the newcomer it's just confusing

snschune committed:
Remove abbreviation
Typo fixes
Requirement change
Update modules/navier_stokes/include/postprocessors/MfrWeightedAverage.h

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/navier_stokes/src/postprocessors/MfrWeightedAverage.C

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/navier_stokes/test/tests/postprocessors/flow_rates/tests

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/navier_stokes/test/tests/postprocessors/flow_rates/tests

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Update modules/navier_stokes/test/tests/postprocessors/flow_rates/tests

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Address comments (#24676)

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/mass_flux_weighted_pp.i:
```diff
@@ -0,0 +1,183 @@
+# Qdot = 500
+# mdot = 1
+# cp = 1
+# dT = 500
```
we're far from that

GiudGiud replied:
we dont quite get that, off by 2e-5 but that s because of the heat source right up to the inlet. 
If we swtich interpolation method, error goes down another factor

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/mass_flux_weighted_pp.i:
```diff
@@ -0,0 +1,183 @@
+# expected temperature at outlet
+#
+# mdot * cp dT = Qdot
+# Qdot = 500
```
we have that one exactly

GiudGiud commented on modules/navier_stokes/test/tests/postprocessors/flow_rates/mass_flux_weighted_pp.i:
```diff
@@ -0,0 +1,183 @@
+
+# expected temperature at outlet
+#
+# mdot * cp dT = Qdot
```
doesnt work?
